generator/sbpg/targets/resources/sbp_tests_template.rs

```diff
@@ -0,0 +1,76 @@
+//
+// Copyright (C) 2015-2018 Swift Navigation Inc.
+// Contact: Swift Navigation <dev@swiftnav.com>
+//
+// This source is subject to the license found in the file 'LICENSE' which must
+// be be distributed together with this source. All other rights reserved.
+//
+// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+// EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+// WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+
+// This file was auto-generated from (((s.src_filename))) by generate.py. Do not modify by hand!
+
+extern crate sbp;
+use sbp::messages::SBPMessage;
+
+trait AlmostEq {
+    fn almost_eq(self, rhs: Self) -> bool;
+}
+
+impl AlmostEq for f32 {
+    fn almost_eq(self, rhs: Self) -> bool {
+        const ULP: f32 = 5.0;
+        ((self - rhs).abs() <= (std::f32::EPSILON * (self + rhs).abs() * ULP)) || ((self - rhs).abs() < std::f32::MIN)
```

Comment:
Not familiar with this formula, is it a standard?

Reply:
Not sure how standard it is, I think I first found in on cppreference (https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon).

`EPSILON` is the distance between `1.0` and the next largest value that can be represented, we then scale it by a value that's on the same order of magnitude as the values (I've seen variations here where there is no scaling done, or the scaling is done by multiplying by the larger of the two values). The idea is that the scaled epsilon value is now about the same size as large as the smallest change in value at the same order of magnitude as the values. We then scale it again by some fudge factor (5 in this case). The final comparison is between the difference between the given values and this fudged smallest representable value.

This isn't always a good way to compare floating point values, but in this case we're trying to verify that the value has been decoded properly from the binary stream (and we aren't performing any mathematical operation on them). I think this comparison against the unit of least precision (ULP) makes sense.

Reply:
cool, thanks for the explanation